p11-kit/rpc-server.c

```diff
@@ -175,18 +176,22 @@ proto_read_ulong_buffer (p11_rpc_message *msg,
 	/* Check that we're supposed to be reading this at this point */
 	assert (!msg->signature || p11_rpc_message_verify_part (msg, "fu"));
 
+	if (!p11_rpc_buffer_get_byte (msg->input, &msg->parsed, &buffer_is_null))
+		return PARSE_ERROR;
 	/* The number of ulongs there's room for on the other end */
 	if (!p11_rpc_buffer_get_uint32 (msg->input, &msg->parsed, &length))
 		return PARSE_ERROR;
 
 	*n_buffer = length;
 	*buffer = NULL;
 
-	/* If set to zero, then they just want the length */
-	if (length == 0)
+	/* If set to zero, then they just want the length for a *buffer = NULL */
+	if (buffer_is_null)
 		return CKR_OK;
 
-	*buffer = p11_rpc_message_alloc_extra (msg, length * sizeof (CK_ULONG));
+	/* Allocate at least 1 element of buf_len even if length is 0 */
```

Comment:
`buf_len` is gone. Also what is the exact reason of doing this? I'm a bit confused because the new test succeeds without this.

Reply:
because we need to allocate at least 1 slot of memory:

- buffer must not be NULL,
- length can be 0


so the underlying OASIS drivers will likely return CKR_BUFFER_TOO_SMALL which means that the lenght MUST be greater than 0 for the next call.

Reply:
It doesn't answer my question, sorry. `p11_rpc_message_alloc_extra` actually allocates at least `sizeof(void *)` bytes even if the argument is 0:
https://github.com/p11-glue/p11-kit/blob/master/p11-kit/rpc-message.c#L102

Could you please show me how it can be a problem, e.g., by writing a unit test reproducing your issue?

Reply:
oh ! I did not notice that `p11_rpc_message_alloc_extra()` allocates at least sizeof(void *), so we are all good.

I'll provide an update of this serie  based on your comments for the 3 cases.

By the way, for my information, why does it always allocate extra memory `(msg->output->frealloc) (NULL, sizeof (void *) + length);` ?